internal/clients/http/ec2/ec2_client.go

```diff
@@ -396,12 +396,18 @@ func (c *ec2Client) RunInstances(ctx context.Context, params *clients.AWSInstanc
 		LaunchTemplate: templateSpec,
 		MaxCount:       ptr.To(amount),
 		MinCount:       ptr.To(amount),
-		InstanceType:   params.InstanceType,
-		ImageId:        ptr.To(params.AMI),
 		KeyName:        &params.KeyName,
 		UserData:       &encodedUserData,
 	}
 
+	if params.InstanceType != "" {
+		input.InstanceType = params.InstanceType
+	}
+
+	if params.AMI != "" {
+		input.ImageId = ptr.To(params.AMI)
+	}
```

Comment:
Okay, this behaviour should be slightly different.
If we do not provide instance type and launch template, this errors out anyway. So, I am not sure if we want to have this case.

> WRN Invalid request: Both instance type and launch template are missing error="instance type or launch template not set"

If we do not provide the ImageId at all (either nothing or a pointer to an empty string), this does not return 400, but ends up with 

> ERR Finishing reservation with error at step 2/3 error="cannot run instances: cannot run instances: operation error EC2: RunInstances, https response error StatusCode: 400, api error MissingParameter: The request must contain the parameter ImageId". But we would create the reservation anyway and return the reservation payload.


Since we can not make a call with an empty ImageId, we should cover this a bit earlier. What do y'all think?

Reply:
First, have you created a launch templates with an image id and with an instance type? We first need to check if the instance is being created from the properties in the template.

Reply:
And that is what I have said- this is an extra mile step, we should check for two cases:
1. Launch template provides instance type/image id and the fields for instance type/image id are empty 

2. Launch templates have the instance type/image id fields propagated and also the fields in the requests are propagated as well. The instance is supposed to be created from the values of the fields not of the template